### ISSUE TITLE
Not able to launch distant on my linode instance

### ISSUE THREAD

**Issue by megalithic** [chipsenkbeil/distant#56]
Hey there Chip!

Huge thanks for this tool. Really looking forward to what it offers for those of us that do our dev work on linode/digitalocean/etc instances. 

I'm having an issue just getting the server to `launch`;

```sh
> distant launch <my_instance_ip> --log-level trace                                                                                                                                        ERROR [distant] Exiting due to error: <my_user>@<my_instance_ip>: Permission denied (publickey).
```

I thought at this point it'd just launch distant for me to then connect to from my local machine (using a pub/priv key auth via ssh).

Are there other things required in my ~/.ssh/sshconfig perhaps?

**Comment by chipsenkbeil** [comment]
Distant spawns ssh to connect to a remote machine, start the distant server, and then exit.

How do you normally connect to your linode instance via ssh? Maybe you need to supply something additional to the ssh command?

Can you share what you run to ssh into the machine? Maybe supply a different identity file?

**Comment by megalithic** [comment]
Thanks for all of this. 

So I've made it a bit further. I went back and rewatched your youtube video again and was able to get things rolling with distant.nvim to at least be able to view session info, but that's it; all other requests fail with `request timed out` (https://github.com/chipsenkbeil/distant.nvim/issues/34). 

I've given it another go with connecting to my instance both in the active process, and as a daemon and here is the ultimate result (showing the interactive run of distant):

```sh
> distant launch <ip> -i ~/.ssh/<priv_key> -u <user>
ERROR [distant] Exiting due to error: Operation timed out (os error 60)
```

**Comment by chipsenkbeil** [comment]
You're seeing this from the launch command? Or subsequent actions? Linode doesn't block ports to my knowledge, so I don't think that would be a problem. This being an actual OS timeout is unexpected.

**Comment by megalithic** [comment]
Here's what i've got for testing this:

`distant launch <ip> -i ~/.ssh/<key>-u <user> --log_file ~/tmp/testdistant.log --log_level trace`

and a result of:

```log
[2021-09-22 21:11:07.053744 -05:00] DEBUG [/Users/<clientuser>/.asdf/installs/rust/stable/registry/src/github.com-1ecc6299db9ec823/distant-0.14.2/src/subcommand/launch.rs:99] Entering interactive loop over unix socket "/var/folders/5n/6y165_gs25xgz6mbyvd059qr0000gn/T/distant.sock"
[2021-09-22 21:11:07.054900 -05:00] DEBUG [/Users/<clientuser>/.asdf/installs/rust/stable/registry/src/github.com-1ecc6299db9ec823/distant-0.14.2/src/subcommand/launch.rs:148] Connecting to <instance_ip> 8088
[2021-09-22 21:11:15.013308 -05:00] ERROR [/Users/<clientuser>/.asdf/installs/rust/stable/registry/src/github.com-1ecc6299db9ec823/distant-0.14.2/src/lib.rs:22] Exiting due to error: Operation timed out (os error 60)

```

**Comment by megalithic** [comment]
for the record; I've not setup anything special here to ssh in to this box; e.g., no fancy sshconfig.

**Comment by chipsenkbeil** [comment]
Okay, so there's ~8 second period before it times out. There's some sort of OS timeout happening and not the software-imposed timeout, which defaults to 15 seconds.

I also noticed that you have other ports taken - maybe by older distant instances - as it's trying to bind to port 8088. You could try killing all of them on the remote server and then starting again. Maybe linode blocks some of those ports.

**Comment by megalithic** [comment]
welp, you nailed it; there were definitly a bunch of other instances running. killed them all and tried it again with the same `launch` configuration above. still hitting the same issue. port 8080 looks open and listening: `tcp        0      0 <ip>:8080      0.0.0.0:*               LISTEN`.

**Comment by megalithic** [comment]
a wee bit more info:

on the server itself, here's my traced logs:

```
[2021-09-22 21:29:31.772057 -05:00] DEBUG [/home/<user>/.asdf/installs/rust/stable/registry/src/github.com-1ecc6299db9ec823/distant-core-0.14.2/src/server/distant/mod.rs:57] Binding to <ip> in range 8080:8099                                                                                                                                   [2021-09-22 21:29:31.772245 -05:00] DEBUG [/home/<user>/.asdf/installs/rust/stable/registry/src/github.com-1ecc6299db9ec823/distant-core-0.14.2/src/server/distant/mod.rs:61] Bound to port: 8080
[2021-09-22 21:30:01.774446 -05:00] DEBUG [/home/<user>/.asdf/installs/rust/stable/registry/src/github.com-1ecc6299db9ec823/distant-core-0.14.2/src/server/utils.rs:80] Shutdown time of 30.000002s has been reached!                                                                                                                                         [2021-09-22 21:30:01.774545 -05:00] WARN [/home/<user>/.asdf/installs/rust/stable/registry/src/github.com-1ecc6299db9ec823/distant-core-0.14.2/src/server/distant/mod.rs:146] Reached shutdown timeout, so terminating
```

**Comment by chipsenkbeil** [comment]
Okay, so no connection is being established with the server at all. Looks like you've configured the server to shutdown after 30 seconds of no active connection, which is a good call to avoid a bunch of attempts cluttering your server. Since no connection is ever received by the server, this implies that either the server is not accessible outside of the linode machine or your local machine is unable to connect to it using the ip and port combo.

You don't have any particular firewall rules on your local machine, do you? Anything that would affect high port ranges? Same for the linode machine.

**Comment by megalithic** [comment]
ahh so more should be happening for a successful connection beyond binding to a port in the given range?

**Comment by chipsenkbeil** [comment]
Right. As soon as a connection is received by the server, you'd see `<Conn @ {}> Established against {}` replacing the curly braces with an id and ip/port combo. Logging is here: https://github.com/chipsenkbeil/distant/blob/master/core/src/server/distant/mod.rs#L117

**Comment by chipsenkbeil** [comment]
@megalithic is this still a blocker for you? If so, you might be interested in #57 once it lands, as it adds support for turning distant into an ssh client that performs all of the operations (file reads, writes, running programs, etc) through the remote ssh server without requiring the distant binary or needing to using a new TCP port as everything is communicated over ssh instead.

The ssh client option is less efficient than having a distant server available, but it could be an alternative for folks who cannot install distant remotely and/or are unable to open a new port on the remote machine.

**Comment by megalithic** [comment]
This is great news! Thanks @chipsenkbeil 

I've still been unable to get this to work. I've ensured ports are accessible via iptables and made sure no vpn or other odd dns things on my client side are happening.

Anyway, looking forward to giving this other way a go. I _really_ wish I could get things running with the way you originally intended via `distant` to `distant` comms.

**Comment by chipsenkbeil** [comment]
@megalithic be on the lookout for https://github.com/chipsenkbeil/distant.nvim/tree/MigrateToRustLib to land. Just now working on it, but it's going to do two things:

1. Remove the need for the distant CLI to be installed on your local machine. Instead, neovim will load the Rust library directly through a `*.so` file.
2. Enable support for using just ssh to do remote operations instead of needing distant on the remote machine. The default will still be distant as it's more performant, stable, and flexible.

**Comment by megalithic** [comment]
@chipsenkbeil I know we briefly touched on this, but, just to confirm, your normal usage of distant/distant.nvim; you have nothing special setup in your ~/.ssh/config for those specific hosts, or (config in general)?

I managed to, one time connect it to a new Digital Ocean instance, and finally as able to get the `:DistantSystemInfo` command to work, buuuut, then not again. ¯\_(ツ)_/¯

**Comment by chipsenkbeil** [comment]
@megalithic 0.15.0 has been released and the support has been merged into distant.nvim as well. You should give it a try and see what happens.

If you are still unable to connect with either distant or ssh mode, then my guess is that you've got something configured on your machine that's causing problems.

**Comment by chipsenkbeil** [comment]
@megalithic going to close this out at the end of the work week if there's nothing else to be done here!
